### PR TITLE
Seperate feedlist

### PR DESCRIPTION
**Pull request by cryptix**
This removes the cpu intensive contacts graph walk from the individual connections and keeps one list in the bot that gets updated periodically.

It also exposes `sbot.Replicate(FeedRef)` methods to fetch fetch without following them publicly and the same for Blocking.

### CODE REVIEW

File: sbot.go [hunk 1]
```diff
@@ -36,6 +36,23 @@ type Indexer interface {
 	GetIndexNamesMultiLog() []string
 }
 
+// Replicator is used to tell the bot which feeds to replicate and which ones to block
```
Comment (freight-stack) [inline sbot.go:39]: Replicate = copy?

File: feedset_test.go [hunk 2]
```diff
@@ -1,22 +1,31 @@
...
 )
 
+func TestFeedSetEmpty(t *testing.T) {
+	r := require.New(t)
+
+	fs := NewFeedSet(0)
+
+	newkey, err := NewKeyPair(nil)
+	r.NoError(err)
+	r.False(fs.Has(newkey.Id))
```
Comment (freight-stack) [inline feedset_test.go:18]: why not require.False(..) and require.NoError(..)?  I have not seen require.New(..) before

File: graph/graph.go [hunk 3]
```diff
@@ -62,22 +62,27 @@ func (g *Graph) Blocks(from, to *ssb.FeedRef) bool {
...
 	for edgs.Next() {
 		nTo := edgs.Node()
 		edg := g.Edge(nFrom.ID(), nTo.ID()).(contactEdge)
 
 		if math.IsInf(edg.Weight(), 1) {
 			ctNode := nTo.(*contactNode)
-			blocked[ctNode.feed.StoredAddr()] = struct{}{}
+			fr, err := ctNode.feed.FeedRef()
+			if err != nil {
+				panic(err)
```
Comment (freight-stack) [inline graph/graph.go:82]: Looks drastic :|

File: graph/graph.go [hunk 4]
```diff
@@ -62,22 +62,27 @@ func (g *Graph) Blocks(from, to *ssb.FeedRef) bool {
...
+			}
+			blocked.AddRef(fr)
+
```
Comment (freight-stack) [inline graph/graph.go:85]: redundant space here

File: plugins/gossip/handler.go [hunk 5]
```diff
@@ -97,18 +96,18 @@ func (g *handler) HandleConnect(ctx context.Context, e muxrpc.Endpoint) {
 		return
 	default:
 	}
-
-	hops := g.GraphBuilder.Hops(g.Id, g.hopCount)
-	if hops != nil {
-		err := g.fetchAll(ctx, e, hops)
+	feeds := g.WantList.ReplicationList()
+	// hops := g.GraphBuilder.Hops(g.Id, g.hopCount)
```
Comment (freight-stack) [inline plugins/gossip/handler.go:100]: Mazbe remove?

File: sbot/new.go [hunk 6]
```diff
@@ -192,12 +198,14 @@ func initSbot(s *Sbot) (*Sbot, error) {
 
 	var inviteService *legacyinvites.Service
 
-	auth := s.GraphBuilder.Authorizer(s.KeyPair.Id, int(s.hopCount+2))
+	// auth := s.GraphBuilder.Authorizer(s.KeyPair.Id, int(s.hopCount+2))
```
Comment (freight-stack) [inline sbot/new.go:201]: Maybe remove?

File: sbot/null_feed_test.go [hunk 7]
```diff
@@ -239,9 +262,30 @@ func TestNullFetched(t *testing.T) {
...
-	r.EqualValues(1000, bobsSeqV.(margaret.Seq).Seq())
+
+	gotMessage := make(chan struct{})
+	updateSink := luigi.FuncSink(func(ctx context.Context, v interface{}, err error) error {
+		seq, ok := v.(margaret.Seq)
+		if !ok {
+			return fmt.Errorf("unexpected type:%T", v)
+		}
+		s := seq.Seq()
+		if s == 999 {
```
Comment (freight-stack) [inline sbot/null_feed_test.go:273]: Why 999? Maybe reference the total count or use a variable to describe its relation?

File: feedset_test.go (hunk 2 above)
Comment (cryptix) [reply feedset_test.go:18]: Mostly because i need to type the full package name plus `t,` :upside_down_face:

File: graph/graph.go (hunk 3 above)
Comment (cryptix) [reply graph/graph.go:82]: i know. the rational is, that only valid feeds are on the graph. I should comment this.

**Pull request closed without merging**